### PR TITLE
do MVV calls with RXJava

### PR DESCRIPTION
**Pull request by pfent**
Follow-up from #1018

Just tested the code, while @kordianbruck just merged and closed the issue :tada: 
Turns out, @thellmund wasn't all wrong when mentioning nobody handled the `JsonParseExceptions`, because we were executing the calls manually only caught `IOExceptions`...
So instead we now do the calls with RxJava, which can properly handle all kinds of errors, so the app shouldn't crash anymore.

I also caught a `IndexOutOfBounds` in `MVVSymbolView`...

### CODE REVIEW

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                    Utils.log(it)
+                    onStartSearch()
+                })
+    }
+
+    private fun displayStations(stations: List<StationResult>) {
+        showLoadingEnded()
+
+        if (stations.isEmpty()) {
+            // So show no results found
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt]: Remove redundant comment

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            listViewResults.adapter = NoResultsAdapter(this)
+            listViewResults.requestFocus()
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt]: Why is that needed?

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            return
+        }
+
+        // mQuery is not null if it was a real search
+        // If there is exactly one station, open results directly
+        if (stations.size == 1 && mQuery != null) {
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationActivity.kt]: Adapt the variable name `mQuery` to Kotlin style (or simply: drop the `m`)

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            listViewResults.requestFocus()
+            return
+        }
+
+        adapterStations.clear()
+        adapterStations.addAll(stations)
+
+        adapterStations.notifyDataSetChanged()
+        listViewResults.adapter = adapterStations
+        listViewResults.requestFocus()
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Why is this needed?

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        setResult(Activity.RESULT_OK, resultValue)
+        finish()
+    }
+
+    /**
+     * Cancel the widget creation and close this activity
+     */
+    private fun cancelAndReturn() {
+        val resultValue = Intent()
+        if (!widgetDepartures.station
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Can you improve the code formatting here?

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            listViewResults.adapter = NoResultsAdapter(this)
+            listViewResults.requestFocus()
+            return
+        }
+
+        adapterStations.clear()
+        adapterStations.addAll(stations)
+
+        adapterStations.notifyDataSetChanged()
+        listViewResults.adapter = adapterStations
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: I think this line is unnecessary. Via the three lines above, the items in the list view should already be updated. Or am I missing something?

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                    Utils.log(it)
+                    onStartSearch()
+                })
+    }
+
+    private fun displayStations(stations: List<StationResult>) {
+        showLoadingEnded()
+
+        if (stations.isEmpty()) {
+            // So show no results found
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Remove redundant comment

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        val stations = TransportController.getRecentStations(recents)
+        displayStations(stations)
+    }
+
+    override fun onStartSearch(query: String) {
+        disposable.add(TransportController.getStationsFromExternal(this, query)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(this::displayStations) {
+                    // Something went wrong
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: We should display something (a `Toast`?) to the user

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            listViewResults.adapter = adapterStations
+            listViewResults.requestFocus()
+        }
+    }
+
+    /**
+     * Click on station in list
+     */
+    override fun onItemClick(av: AdapterView<*>, v: View, position: Int, id: Long) {
+        val (station, id1) = av.adapter.getItem(position) as StationResult
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Please use a different variable name

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+        val tm = TransportController(this)
+        this.widgetDepartures = tm.getWidget(appWidgetId)
+
+        val autoReloadSwitch = findViewById<Switch>(R.id.mvv_widget_auto_reload)
+        autoReloadSwitch.isChecked = this.widgetDepartures.autoReload
+        autoReloadSwitch.setOnCheckedChangeListener { _, checked ->
+            widgetDepartures.autoReload = checked
+        }
+        // TODO add handling for use location
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Was this TODO in here before you did the refactoring?

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    private fun loadDetails(location: String, locationID: String) {
+        // Quality is always 100% hit
+        val stationResult = StationResult(location, locationID, Integer.MAX_VALUE)
+        val jsonStationResult = gson.toJson(stationResult)
+
+        // save clicked station into db
+        recentsDao.insert(Recent(jsonStationResult, RecentsDao.STATIONS))
+
+        // Check for internet connectivity
+        if (!NetUtils.isConnected(this)) {
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt]: This check is performed in the `ConnectivityInterceptor`. You can remove it here.

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+     * @param results List of departures
+     */
+    private fun displayResults(results: List<Departure>?) {
+        showLoadingEnded()
+        if (results == null || results.isEmpty()) {
+            showError(R.string.no_departures_found)
+            return
+        }
+        mViewResults.removeAllViews()
+        for ((_, direction, symbol1, _, departureTime) in results) {
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt]: Please rename `symbol1`.

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        return true
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        if (item.itemId == R.id.action_transport_usage) {
+            AlertDialog.Builder(this)
+                    .setTitle(R.string.transport_action_usage)
+                    .setMessage(R.string.transport_help_text)
+                    .setPositiveButton(android.R.string.ok, null)
+                    .create()
```
Comment (thellmund) [inline app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt]: You can call `show()` on the builder itself

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt (hunk 10 above)
Comment (pfent) [reply app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidgetConfigureActivity.kt]: Yes, that was there before refactoring...
